src/V3Ast.h

```diff
@@ -1921,11 +1931,110 @@ class AstNode VL_NOT_FINAL {
+    // Traverse subtree and call given function 'f' in pre-order on each node that has type 'T'.
+    // Prefer 'foreach' over simple VNVisitor that only needs to handle a single (or a few) node
+    // types, as it's easier to write, but more importantly, the dispatch to the operation function
+    // in 'foreach' should be completely predictable by branch target caches in modern CPUs,
+    // while it is basically unpredictable for VNVisitor.
+    template <typename T_Node> void foreach (std::function<void(T_Node*)> f) {
+        static_assert(!std::is_const<T_Node>::value,
+                      "Type parameter 'T_Node' should not be const qualified");
+        static_assert(std::is_base_of<AstNode, T_Node>::value,
+                      "Type parameter 'T_Node' must be a subtype of AstNode");
+        foreachImpl<T_Node, /* VisitNext: */ false>(this, f);
+    }
+
+    // Same as above, but for 'const' nodes
+    template <typename T_Node> void foreach (std::function<void(const T_Node*)> f) const {
+        static_assert(!std::is_const<T_Node>::value,
+                      "Type parameter 'T_Node' should not be const qualified");
+        static_assert(std::is_base_of<AstNode, T_Node>::value,
+                      "Type parameter 'T_Node' must be a subtype of AstNode");
+        foreachImpl<const T_Node, /* VisitNext: */ false>(this, f);
+    }
+
+    // Same as 'foreach' but also follows 'this->nextp()'
+    template <typename T_Node> void foreachAndNext(std::function<void(T_Node*)> f) {
+        static_assert(!std::is_const<T_Node>::value,
+                      "Type parameter 'T_Node' should not be const qualified");
+        static_assert(std::is_base_of<AstNode, T_Node>::value,
+                      "Type parameter 'T_Node' must be a subtype of AstNode");
+        foreachImpl<T_Node, /* VisitNext: */ true>(this, f);
+    }
+
+    // Same as 'foreach' but also follows 'this->nextp()'
+    template <typename T_Node> void foreachAndNext(std::function<void(const T_Node*)> f) const {
+        static_assert(!std::is_const<T_Node>::value,
+                      "Type parameter 'T_Node' should not be const qualified");
+        static_assert(std::is_base_of<AstNode, T_Node>::value,
+                      "Type parameter 'T_Node' must be a subtype of AstNode");
+        foreachImpl<const T_Node, /* VisitNext: */ true>(this, f);
+    }
+
+    size_t nodeCount() const {
+        size_t count = 0;
+        this->foreach<AstNode>([&count](const AstNode*) { ++count; });
+        return count;
+    }
 };
 
-// Specialisations of privateIs/privateCast
+// Specialisations of privateTypeTest
 #include "V3Ast__gen_impl.h"  // From ./astgen
 
+// Specializations of privateMayBeUnder
+template <> inline bool AstNode::mayBeUnder<AstCell>(const AstNode* nodep) {
+    return !VN_IS(nodep, NodeStmt) && !VN_IS(nodep, NodeMath);
+}
+template <> inline bool AstNode::mayBeUnder<AstNodeAssign>(const AstNode* nodep) {
+    return !VN_IS(nodep, NodeMath);
+}
+
```

Comment:
Ideally I think we want this in the AstCell etc class definition, then ast gen to process something to make these, so it's closer to where people want to read it.  I'm ok waiting to do that.

Reply:
Agreed, I would go one further and strengthen the accessors/constructors so you can prove these relationships statically, which astgen can then extract.

Reply:
BTW proving statically will nicely make sure "new" works, but most of the hard bugs are due to bad edits, which don't call the accessors/constructors (e.g. lots of replaceWith's), so we must always have V3Broken do a dynamic check regardless.